demo/ota-agent/main.c

```diff
@@ -0,0 +1,236 @@
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "FreeRTOS.h"
+#include "semphr.h"
+#include "task.h"
+
+#include "core_mqtt.h"
+#include "mqtt_wrapper.h"
+#include "ota_demo.h"
+#include "transport/transport_wrapper.h"
+#include "utils/clock.h"
+
+#define MAX_THING_NAME_SIZE 128U
+
+static TransportInterface_t transport = { 0 };
+static MQTTContext_t mqttContext = { 0 };
+static uint8_t networkBuffer[ 5000U ];
+
+static StaticSemaphore_t MQTTAgentLockBuffer;
+static StaticSemaphore_t MQTTStateUpdateLockBuffer;
+SemaphoreHandle_t MQTTAgentLock = NULL;
+SemaphoreHandle_t MQTTStateUpdateLock = NULL;
+
+static void otaAgentTask( void * parameters );
+
+static void mqttProcessLoopTask( void * parameters );
+
+static void suspendResumeLoopTask( void * parameters );
+
+static void mqttEventCallback( MQTTContext_t * mqttContext,
+                               MQTTPacketInfo_t * packetInfo,
+                               MQTTDeserializedInfo_t * deserializedInfo );
+
+static void handleIncomingMQTTMessage( char * topic,
+                                       size_t topicLength,
+                                       uint8_t * message,
+                                       size_t messageLength );
+
+int main( int argc, char * argv[] )
+{
+    MQTTStatus_t mqttResult;
+    MQTTFixedBuffer_t fixedBuffer = { 0 };
+
+    if( argc != 6 )
+    {
+        printf( "Usage: %s certificateFilePath privateKeyFilePath "
+                "rootCAFilePath endpoint thingName\n",
+                argv[ 0 ] );
+        return 1;
+    }
+
+    MQTTAgentLock = xSemaphoreCreateRecursiveMutexStatic(
+        &MQTTAgentLockBuffer );
+    MQTTStateUpdateLock = xSemaphoreCreateMutexStatic(
+        &MQTTStateUpdateLockBuffer );
+
+    fixedBuffer.pBuffer = networkBuffer;
+    fixedBuffer.size = 5000U;
+
+    transport_tlsInit( &transport );
+    mqttResult = MQTT_Init( &mqttContext,
+                            &transport,
+                            Clock_GetTimeMs,
+                            mqttEventCallback,
+                            &fixedBuffer );
+    assert( mqttResult == MQTTSuccess );
+
+    xTaskCreate( otaAgentTask, "T_OTA", 6000, ( void * ) argv, 1, NULL );
+    xTaskCreate( mqttProcessLoopTask, "T_MQTT", 6000, NULL, 2, NULL );
+    xTaskCreate( suspendResumeLoopTask, "T_SUSPEND", 6000, NULL, 2, NULL );
+
+    mqttWrapper_setCoreMqttContext( &mqttContext );
+    mqttWrapper_setThingName( argv[ 5 ],
+                              strnlen( argv[ 5 ], MAX_THING_NAME_SIZE ) );
+
+    vTaskStartScheduler();
+
+    return 0;
+}
+
+static void suspendResumeLoopTask( void * parameters )
+{
+    ( void ) parameters;
+    OtaEventMsg_t testEvent = { 0 };
+    OtaState_t curState;
+    printf("Suspend resume task \n");
+    while( true )
+    {
+        curState = getOtaAgentState();
+
+        if  (curState != OtaAgentStateStopped && curState >= OtaAgentStateRequestingJob && curState != OtaAgentStateSuspended)
```

Comment:
Why not `if (curState >= OtaAgentStateRequestingJob && curState <= OtaAgentStateSuspended)`? I think we'd want to respect the shutting down state

Reply:
Fixed this in the newer version.